source/common/protobuf/message_converter.h

```diff
@@ -0,0 +1,113 @@
+//
+// MessageConverter is thread-compatible, but it is not tested in multi-threaded
+// environments. This class is meant to be used in Envoy's data plane, where
+// we will have 1 separate MessageConverter instance per worker thread.
+//
+
+class MessageConverter {
+public:
+  MessageConverter(std::unique_ptr<CreateMessageDataFunc> factory)
+      : MessageConverter(std::move(factory), std::numeric_limits<uint32_t>::max()) {}
+
+  explicit MessageConverter(std::unique_ptr<CreateMessageDataFunc> factory, uint32_t buffer_limit);
+
+  // Accumulates a gRPC data buffer into a single StreamMessage.
+  //
+  // Caller can expect all data to be moved out of the input `data` and
+  // potentially buffered internally in this `MessageConverter` object.
+  //
+  // Returns OK with nullptr if only an incomplete message exists
+  //         OK with the message, when all fragments of a message were collected
+  //         OK with the message and nullptr data, when it is the last message
+  //         in a stream
+  //         Otherwise, an error indicating the failure to create a message.
+  absl::StatusOr<std::unique_ptr<StreamMessage>> AccumulateMessage(Envoy::Buffer::Instance& data,
+                                                                   bool end_stream);
+
+  // The same as `AccumulateMessage`, but returns a list of `StreamMessage`.
+  // This should be used if the caller supports streaming, as one data buffer
+  // maybe have multiple gRPC frames embedded.
+  //
+  // Caller is expected to call `ConvertBackToBuffer` in the same order
+  // these StreamMessages are returned.
+  //
+  // Returns OK with `n-1` messages guaranteed to have data. Those messages
+  //         were fully parsed. The last message may be a valid message with
+  //         nullptr data if it is the last message in the stream.
+  //         Otherwise, an error indicating the failure to create a message.
+  absl::StatusOr<std::vector<std::unique_ptr<StreamMessage>>>
+  AccumulateMessages(Envoy::Buffer::Instance& data, bool end_stream);
+
+  // Converts the StreamMessage back into an Envoy data buffer.
+  //
+  // Ownership of `message` is passed to this function.
+  // Once conversion occurs, `message` is no longer valid and must NOT be used.
+  //
+  // Returns an Envoy Buffer with the same data contents as the `message.`
+  // Returns a status if conversion failed.
+  absl::StatusOr<Envoy::Buffer::InstancePtr>
+  ConvertBackToBuffer(std::unique_ptr<StreamMessage> message);
+
+  // Returns the total number of bytes stored across all internal buffers.
+  // Callers are not expected to explicitly check this, as we internally ensure
+  // we do not go over buffer limits. But callers may use this for their own
+  // watermarking.
+  [[nodiscard]] uint64_t BytesBuffered() const;
+
+  // MessageConverter is neither copyable nor movable.
+  MessageConverter(const MessageConverter&) = delete;
+  MessageConverter& operator=(const MessageConverter&) = delete;
```

Comment:
nit: These should go up to be just after the constructors.

Reply:
Done.